### DISCUSSION TITLE
Next.js 14 Setup must use Store Provider?

### DISCUSSION THREAD

**Discussion by emiltsang-flow** [pmndrs/zustand#2426] [Q&A]
I am developing a project with `Next.js 14` with `Zustand` as global state management.

From the doc, https://docs.pmnd.rs/zustand/guides/nextjs,
i found there is a setup guide with the `store provider pattern`.

If i follow the rules that `ONLY` use zustand in `client components`,
is it a `MUST` to use `store provider` for every store setup?

**Comment by dbritto-dev** [comment]
@emiltsang-flow it's

**Reply by emiltsang-flow** [reply]
so it is required to implement with store provider?
and how should i wrap the `children in layout.tsx` if i have `multiple stores`?

**Reply by dbritto-dev** [reply]
@emiltsang-flow that's a good question currently we don't have a section for that but I can add it to the nextjs guide

**Reply by dbritto-dev** [reply]
But I would use the same pattern for multiple providers something like this:

```
<StoreProviders providers={[BearStoreProvider, FishStoreProvider, ...]}>
  {...}
</StoreProviders>
```

**Comment by dai-shi** [comment]
> If i follow the rules that ONLY use zustand in client components

It's a subtle difference, but if you ONLY use zustand on the client, a global store is fine. There're some techniques like `<ClientOnly>`. https://www.joshwcomeau.com/react/the-perils-of-rehydration/#abstractions
But, in general, people don't know that technical detail and even experienced developers could miss it, so using Store Provider would always be better.

> sounds a bit complicated though

So, no, using something only on the client is more complicated.

**Reply by emiltsang-flow** [reply]
@dai-shi 
how should i wrap the children in `layout.tsx` if i have `multiple zustand stores`?
any methods to create `ONLY 1 store provider` component throughout the app?

**Reply by dai-shi** [reply]
I don't know. At that point, I don't think it makes sense to use nextjs anyway.

**Reply by dai-shi** [reply]
Ah, I think I misunderstood your question. It has nothing to do with my previous comment.

**Comment by abelta** [comment]
This is actually a good question. From reading the documentation, it seems that kind of setup is mandatory but Zustand works fine in my projects and other people's without it. I can't even find any articles where such configuration is used.

Reading this thread I'm finding out that the particular setup is only necessary if Zustand is going to be used in the server which I don't even think is the preferred use. Maybe this should be made more clear in the docs?

**Reply by dbritto-dev** [reply]
@abelta you are misunderstading some concepts but here's an explanation about that -> https://github.com/pmndrs/zustand/discussions/2326#discussioncomment-8531033

> **Note:** you can safely use zustand on nextjs by using client only components or by using `next/dynamic` and disabling ssr

**Reply by dbritto-dev** [reply]
btw, these concepts are applied to redux too since redux and zustand are global stores and/or module states

**Comment by lsbyerley** [comment]
I'm seeing a `You're importing a component that needs useRef. It only works in a Client Component but none of its parents are marked with "use client", so they're Server Components by default.` when following the setup for the App Router method in the Zustand docs. It looks like layout.js is rendering the provider with uses client side methods (https://docs.pmnd.rs/zustand/guides/nextjs#app-router). This seems incorrect?

**Reply by dbritto-dev** [reply]
@lsbyerley I don't get it. What's your question?

**Reply by lsbyerley** [reply]
The `src/app/layout.tsx` file example [in the app router example](https://docs.pmnd.rs/zustand/guides/nextjs#app-router) is generating an error because it's importing the provider which uses the `useRef` client side hook.

**Reply by dbritto-dev** [reply]
@lsbyerley that's wrong `useRef` is not a client hook and shouldn't trigger any error

**Reply by lsbyerley** [reply]
I am seeing `You're importing a component that needs createContext. It only works in a Client Component but none of its parents are marked with "use client", so they're Server Components by default.`

and 

`You're importing a component that needs useRef. It only works in a Client Component but none of its parents are marked with "use client", so they're Server Components by default.`

both pointing to the `@/providers/counter-store-provider` file in the nextjs app router example. Do you have a codesandbox link with this working? Just trying to understand why

**Reply by dbritto-dev** [reply]
@lsbyerley sure, would you mind sharing me your implementation too?

**Reply by lsbyerley** [reply]
@dbritto-dev thank you for the back and forth. I've found the issue and it was a stupid typo on my end. Thanks again!

**Reply by dbritto-dev** [reply]
@lsbyerley btw, here's the demo -> https://github.com/dbritto-dev/zustand-next-demo
